core/consensus/grandpa/impl/authority_manager_impl.cpp

```diff
@@ -0,0 +1,1224 @@
+
+    if (node->block == block) {
+      node->adjust(node_in_finalized_chain);
+      OUTCOME_TRY(disable_authority(node));
+    } else {
+      auto new_node = node->makeDescendant(block, node_in_finalized_chain);
+
+      OUTCOME_TRY(disable_authority(new_node));
+
+      // Reorganize ancestry
+      auto descendants = std::move(node->descendants);
+      for (auto &descendant : descendants) {
+        if (directChainExists(block, descendant->block)) {
+          // Propagate change to descendants
+          if (descendant->authorities == node->authorities) {
+            descendant->authorities = new_node->authorities;
+          }
+          new_node->descendants.emplace_back(std::move(descendant));
+        } else {
+          node->descendants.emplace_back(std::move(descendant));
+        }
+      }
+      node->descendants.emplace_back(std::move(new_node));
+    }
+
+    return outcome::success();
+  }
+
+  outcome::result<void> AuthorityManagerImpl::applyPause(
+      const primitives::BlockInfo &block, primitives::BlockNumber activate_at) {
+    SL_DEBUG(logger_, "Applying pause on block {}", block);
+
+    auto node = getNode(block);
+
+    if (not node) {
+      return AuthorityManagerError::ORPHAN_BLOCK_OR_ALREADY_FINALIZED;
+    }
+
+    IsBlockFinalized node_in_finalized_chain =
+        node->block.number <= block_tree_->getLastFinalized().number;
+
+    auto new_node = node->makeDescendant(block, node_in_finalized_chain);
+
+    new_node->action = ScheduleNode::Pause{activate_at};
+
+    SL_VERBOSE(
+        logger_, "Scheduled pause after block #{}", new_node->block.number);
+
+    // Reorganize ancestry
+    auto descendants = std::move(node->descendants);
+    for (auto &descendant : descendants) {
+      auto &ancestor =
+          block.number <= descendant->block.number ? new_node : node;
+      ancestor->descendants.emplace_back(std::move(descendant));
+    }
+    node->descendants.emplace_back(std::move(new_node));
+
+    return outcome::success();
+  }
+
+  outcome::result<void> AuthorityManagerImpl::applyResume(
+      const primitives::BlockInfo &block, primitives::BlockNumber activate_at) {
+    auto node = getNode(block);
+
+    if (not node) {
+      return AuthorityManagerError::ORPHAN_BLOCK_OR_ALREADY_FINALIZED;
+    }
+
+    IsBlockFinalized node_in_finalized_chain =
+        node->block.number <= block_tree_->getLastFinalized().number;
+
+    auto new_node = node->makeDescendant(block, node_in_finalized_chain);
+
+    new_node->action = ScheduleNode::Resume{activate_at};
+
+    SL_VERBOSE(
+        logger_, "Resuming will be done at block #{}", new_node->block.number);
+
+    // Reorganize ancestry
+    reorganize(node, new_node);
+
+    return outcome::success();
+  }
+
+  outcome::result<void> AuthorityManagerImpl::onDigest(
+      const primitives::BlockInfo &block,
+      const consensus::babe::BabeBlockHeader &digest) {
+    auto node = getNode(block);
```

Comment:
getNode should be renamed, it returns ancestor block AFAIU

Reply:
It returns the matching node, or the best ancestor if there is no matching node. I'll leave neutral name **getNode**